### ISSUE TITLE
Connection receive_timeout regression when updating from 2.2.4 to 2.4

### ISSUE THREAD

**Issue by jgehrcke** [cannatag/ldap3#438]
We updated from ldap3 2.2.4 to ldap3 2.4 and one of our integration tests caught a regression around the `receive_timeout` argument of `Connection` objects.

In our application, we create a `Connection` object with a `receive_timeout` argument of 7 seconds. Later we call `conn.start_tls()`. In our integration test, this expected to error our not earlier than after 7 seconds. With ldap3 2.2.4 this test passes.

After updating to ldap3 2.4 our integration test fails: the timeout error is actually raised after roughly 1 second.

I saw https://github.com/cannatag/ldap3/issues/356 and did a bit of debugging with the following patch to `strategies/sync.py`:

```
$ diff -wu sync.py.orig sync.py
--- sync.py.orig	2017-11-23 18:16:40.536759062 +0100
+++ sync.py	2017-11-23 18:16:14.274747513 +0100
@@ -80,8 +80,18 @@
         while receiving:
             if get_more_data:
                 try:
+
+                    import socket
+                    print('socket option SO_RCVTIMEO:')
+                    print(self.connection.socket.getsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO))
+                    print('socket.gettimeout:')
+                    print(self.connection.socket.gettimeout())
+
                     data = self.connection.socket.recv(self.socket_size)
                 except (OSError, socket.error, AttributeError) as e:
+                    print(str(e))
+                    print(type(e))
+
                     self.connection.last_error = 'error receiving data: ' + str(e)
                     exc = e
```
This prints
```
socket option SO_RCVTIMEO:
7
socket.gettimeout:
1.0
timed out
<class 'socket.timeout'>
```
That is, the `SO_RCVTIMEO` option has been set as expected. However, `socket.gettimeout()` is the method corresponding to `socket.settimeout()`, and it returns 1 instead of 7. `socket.settimeout()` seems to be the only "official way" to actually set a timeout value from CPython's `socket` module. One important aspect to check next is whether or not ldap3 2.4 actually uses the `socket.settimeout()` method for setting the timeout.

Note that the `Server`'s `connect_timeout` is set to 1 s. 

This has been observed with CPython 3.6.3.

**Comment by jgehrcke** [comment]
It looks like ldap3 is making mixed usage of using `settimeout()` and setting `SO_RCVTIMEO` directly:
```
$ grep -nR 'settimeout'
core/server.py:296:                    temp_socket.settimeout(self.connect_timeout)
core/server.py:298:                    temp_socket.settimeout(conf_availability_timeout)  # set timeout for checking availability to default
strategy/base.py:207:                self.connection.socket.settimeout(self.connection.receive_timeout)
strategy/base.py:223:                self.connection.socket.settimeout(self.connection.server.connect_timeout)
strategy/base.py:249:            self.connection.socket.settimeout(None)  # disable socket connection timeout - socket is in blocking mode or in unblocking mode if receive_timeout is specified in connection
```

```
$ grep -nR 'SO_RCVTIMEO'
strategy/base.py:209:                    self.connection.socket.setsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO, int(1000 * self.connection.receive_timeout))
strategy/base.py:211:                    self.connection.socket.setsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO, pack('LL', self.connection.receive_timeout, 0))
```

**Comment by jgehrcke** [comment]
On a modern Python it's definitely not sufficient to only set `SO_RCVTIMEO` when calling Python's `recv()`:

- `sock_recv_guts()` uses `sock_call()`: https://github.com/python/cpython/blob/b98f1715a35d2cbd1e9e45e1e7ae51a39e00dc4a/Modules/socketmodule.c#L3146
- `sock_call()` uses `sock_call_ex()` which implements its own timeout logic using non-blocking calls: https://github.com/python/cpython/blob/b98f1715a35d2cbd1e9e45e1e7ae51a39e00dc4a/Modules/socketmodule.c#L751 ("The connection runs asynchronously.")

**Comment by cannatag** [comment]
Thanks Jan-Philip, will work on this issue in the weekend.

Bye,
Giovanni

**Comment by jgehrcke** [comment]
The PR I have submitted for addressing this issue (I have confirmed that it makes our integration test pass again) is https://github.com/cannatag/ldap3/pull/439. I'd appreciate your review!

**Comment by jgehrcke** [comment]
I want to add that while working on the patch the root cause of the issue became obvious in more detail: the problem only arises when one uses a connect timeout _plus_ a receive timeout, and when the connect timeout is _smaller than_ the receive timeout.

**Comment by cannatag** [comment]
Hi, I've accepted your PR, can you check the code in dev?

**Comment by cannatag** [comment]
Released in version 2.4.1
